kvbc/src/kvbc_app_filter/kvbc_app_filter.cpp

```diff
@@ -418,24 +418,21 @@ FindGlobalEgIdResult KvbAppFilter::findGlobalEventGroupId(uint64_t external_even
   // At this point, we exhausted all private entries => it has to be a public event group
   uint64_t global_eg_id;
 
+  // If all private event groups were pruned then we need to adjust current_* to point to the last private event group
+  if (private_start == 0) {
+    ConcordAssertNE(private_end, 0);
+    current_pvt_eg_id = private_end;
+    current_ext_eg_id = private_end + public_start - 1;
+  }
+
   if (external_event_group_id < current_ext_eg_id) {
-    if (current_pvt_eg_id == private_start) {
```

Comment:
Isn't this condition still a valid optimization for non-pruning scenarios, i.e., if `private_start != 0`?

Reply:
This part of the code is actually wrong and the pruning scenarios unveiled it. Just because our current pointer points to private_start doesn't mean that there were no private event groups before that point. In fact, if we pruned partially then there will be private event groups before private_start but they aren't queryable. Therefore, we cannot simply use the given external event group id but instead need to do proper calculation.
If we do that right here then we just duplicating the same logic for the "non-optimized" case and with that we end up similar to the code below which is redundant.